src/advanced/time.md

```diff
@@ -0,0 +1,287 @@
+  including during nodes update.
+- **Sufficient accuracy**. The specified time should be fairly accurate.
+  In practice, an acceptable deviation is a few seconds.
+- **Monotony**. The time value should only increase. A pragmatic requirement,
+  which simplifies use of time when implementing the business logic.
+
+## Preliminaries
+
+Two possible approaches satisfying the above criteria are
+[integration with consensus](#integration-with-consensus)
+and [a time oracle service](#time-oracle-service).
+
+### Integration with Consensus
+
+!!! tip
+    Consult the [consensus description](../architecture/consensus.md) for
+    explanation of `Propose` and `Precommit` messages, and the “+2/3” notation.
+
+The validator includes its local time in each `Precommit` message.
+At the next height the leader includes +2/3 `Precommit`s of the previous block
+into the `Propose`. The time median of these `Precommit`s is recorded into
+the header of the block obtained based on this `Propose`.
+
+#### Advantages
+
+- The time value is indicated directly in the header of each block making it
+  more accessible.
+- Time is forcibly updated in the course of consensus operation:
+  it is impossible to sabotage update of time without stopping consensus.
+- Blockchain is not clogged by transactions associated with time determination.
+
+#### Disadvantages
+
+- The consensus code becomes more complex. (Time is included into the consensus
+  logic while [anchoring](bitcoin-anchoring.md) and
+  [configuration](configuration-updater.md) are not.)
+- Time updates are tied to creation of `Precommit` messages. In the case
+  of a large delay in block acceptance, all the transactions therein will
+  be executed with the same outdated time value.
+
+### Time Oracle Service
+
+Each validator at a specific time sends a transaction indicating its local time
+(usually immediately after the commit of each block). Exonum storage contains
+an index with the most current time indicated separately by each validator.
+A median of these index values is stored separately; it is considered the
+actual time and is updated after each transaction from any of the validators.
+
+#### Advantages
+
+- The logic for time update is placed in a separate plug-in service
+  (modularity).
+- In case of a long delay in block acceptance, the time will be updated along
+  with the delayed block execution while executing its transactions.
+  The time will be accurate enough with regard to the time of the transaction
+  entry into the mempool.
+
+#### Disadvantages
+
+- The time value is stored in the Exonum storage; hence,
+  verifying its authenticity requires additional
+  cryptographic checks from lightweight clients.
+- Exonum blocks are burdened with time oracle transactions
+  (usually one for each validator).
+
+### Choosing Approach
+
+Given the pros and cons above, **exonum-time** uses the time oracle approach
+as a more flexible one.
```

Comment:
I might misunderstand the purpose of this document, but is it even necessary to put descriptions of various approaches and such design decisions in the user guide? As a user I want to know how to use the time oracle and what to expect from it. As an inquisitive user, I might be interested in how it works. But design decisions are beyond that interest. I believe that is redundant information in the user guide, but it might be useful as project documentation.

Reply:
I think it could be worth it to reposition the "Preliminaries" section as "Possible alternatives", where the alternative (i.e., integration with consensus) would be described. This would be in line with academic research, which the article already emulates to an extent.